### PR TITLE
pipeline-manager: add a config API endpoint to get a list of canned demo URLs

### PR DESCRIPTION
**Pull request by lalithsuresh**
Signed-off-by: Lalith Suresh <lalith@feldera.com>

Is this a user-visible change (yes/no): yes

### CODE REVIEW

**Comment by snkas** [comment]
* Canned demo endpoint now returns a list of JSON objects of type:
  ```
  [
    {
        "title": "...",
        "description": "...",
        "url": "..."
    },
    ...
  ]
  ```
 
* Added a first canned demo: warehouse
* Rebased against main (kept the Cargo.lock changes)
* Minor comment and text changes

**Comment by snkas** [comment]
Command to run it directly:

```
cargo run --bin=pipeline-manager --features pg-embed -- --dev-mode --canned-demos='{"title": "Warehouse", "description": "Use case of a warehouse that stores product items which are bought and sold.", "url": "https://raw.githubusercontent.com/feldera/feldera/features/demo/demos/warehouse/demo.json"}'
```

... and then you can check the endpoint using:

```
curl -X GET 'http://localhost:8080/v0/config/demos' | jq
```

**Comment by lalithsuresh** [comment]
@snkas My understanding was that the title and description will be encoded in the json file that describes the demo. I don't think we need to repeat that in the CLI argument here. Besides, it'll be a pain to render via Helm with all the escaping. :)

**Comment by snkas** [comment]
@lalithsuresh , @Karakatiza666 had a strong preference to have the name and description be part of the demo list itself I think for UI placeholders while the demo JSON is loaded in. I agree it becomes a pain in terms of escaping when passing serialized JSON... We can keep it a list of URLs in this first version and then revisit it later on. What do you think @Karakatiza666 ?

**Comment by snkas** [comment]
Reverted to just a list of strings (URLs).

**Comment by snkas** [comment]
Split off the canned demo warehouse to separate PR: https://github.com/feldera/feldera/pull/1401

Command to run it directly:

```
cargo run --bin=pipeline-manager --features pg-embed -- --dev-mode --canned-demos=https://raw.githubusercontent.com/feldera/feldera/canned-demo-warehouse/demo/demos/warehouse/demo.json
```

... and then you can check the endpoint using:

```
curl -X GET 'http://localhost:8080/v0/config/demos' | jq
```

**Comment by snkas** [comment]
Ready for final review
- Didn't add any demo to docker-compose.yml (yet), because: (1) `test-docker-compose-stable` in Earthly causes the previous image to get called with an argument (`--canned-demos ..`) it does not know which causes CI to fail, and (2) there is not yet a canned demo available besides in PR https://github.com/feldera/feldera/pull/1401 , which will be merged later.
  Not entirely sure what would be the best way to solve issue (1) to get a passing CI. (2) can be solved by merging it earlier or hosting the demo files somewhere else.
- Added CHANGELOG entry
- @lalithsuresh : I kept the Cargo.lock changes from your original commit

File: crates/pipeline_manager/src/api/config_api.rs [hunk 1]
```diff
@@ -34,5 +34,23 @@ async fn get_authentication_config(
     Ok(HttpResponse::Ok().json(&auth_config.provider))
 }
 
+/// Get the list of canned demo URLs.
+#[utoipa::path(
+    path="/config/demos",
+    responses(
+        (status = OK
+            , description = "URLs to JSON objects that self-describe a set of canned demos"
```
Comment (ryzhyk) [inline crates/pipeline_manager/src/api/config_api.rs:42]:
```suggestion
            , description = "URLs to JSON objects that describe a set of canned demos"
```

File: crates/pipeline_manager/src/api/config_api.rs [hunk 2]
```diff
@@ -34,5 +34,23 @@ async fn get_authentication_config(
...
+            , content_type = "application/json"
+            , body = Vec<String>),
+    ),
+    context_path = "/v0",
+    security(("JSON web token (JWT) or API key" = [])),
+    tag = "Configuration",
+)]
+#[get("/config/demos")]
```
Comment (ryzhyk) [inline crates/pipeline_manager/src/api/config_api.rs:50]: Does this have to be under `/config`? Seems like demos should be a separate tag.

**Review by ryzhyk — APPROVED** [review]

File: crates/pipeline_manager/src/api/config_api.rs (hunk 2 above)
Comment (snkas) [reply crates/pipeline_manager/src/api/config_api.rs:50]: The reasoning is I think that it is part of the configuration with which you start the pipeline-manager (it is provided as a command-line argument), as such you fetch it from the config.

File: crates/pipeline_manager/src/api/config_api.rs (hunk 1 above)
Comment (snkas) [reply crates/pipeline_manager/src/api/config_api.rs]: Changed :+1:

File: crates/pipeline_manager/src/config.rs [hunk 3]
```diff
@@ -163,6 +163,13 @@ pub struct ApiServerConfig {
     #[serde(default)]
     #[arg(long)]
     pub dev_mode: bool,
+
+    /// Used for supplying clients like the UI with canned demos. Administrators
+    /// can use this option to set up environment-specific demos for users
+    /// (e.g., ones that connect to an internal data source).
+    #[serde(default)]
+    #[arg(long)]
+    pub canned_demos: Vec<String>,
```
Comment (gz) [inline crates/pipeline_manager/src/config.rs:172]: what's the point of adding the 'canned', why not just call it 'demos'

File: crates/pipeline_manager/src/api/config_api.rs (hunk 2 above)
Comment (ryzhyk) [reply crates/pipeline_manager/src/api/config_api.rs:50]: Yeah, I get it. I do think it's a bit of a stretch, but I'm ok with it.

**Comment by snkas** [comment]
Rebase with main, awaiting CI to pass to merge.

**Pull request merged**